### PR TITLE
feat: Update logic for ShardToNodeRatio and MinIndexReplicas

### PR DESCRIPTION
**Pull request by s-vkropotko**
## Description
With this PR I try to fix a bug with setting minIndexReplicas on config change and update the logic for a ShardToNodeRatio param. 
Fix a function calculateNodesWithSameShardToNodeRatio, before if we have shards < nodes, the function conducts division to the number what is <1 (0 < n < 1), as a result function increases replicas.  
About an evaluation of number of nodes on checking MinShardsPerNode: looks like it makes sense to do it only in case of the new number of shards is more than MinShardsPerNode. 

## Types of Changes
Fix bug #141
Minor changes for ShardToNodeRatio param logic

## Review
_List of tasks the reviewer must do to review the PR_
- [x] Tests

## Deployment Notes
These should highlight any db migrations, feature toggles, etc.

### CODE REVIEW

**Comment by mikkeloscar** [comment]
@s-vkropotko Thanks! Could you add a test covering this change

**Comment by s-vkropotko** [comment]
@mikkeloscar hey! Could you please check tests in the PR?

File: operator/autoscaler_test.go [hunk 1]
```diff
@@ -272,8 +272,8 @@ func TestIncreaseShardToNodeRatioMore(t *testing.T) {
 
 	// scale-down even if this means increasing shard-to-node ratio of more than +1
 	esIndices := map[string]ESIndex{
-		"ad1": {Replicas: 0, Primaries: 21, Index: "ad1"},
-		"ad2": {Replicas: 0, Primaries: 2, Index: "ad2"},
+		"ad1": {Replicas: 1, Primaries: 21, Index: "ad1"},
```
Comment (otrosien) [inline operator/autoscaler_test.go:275]: I prefer adding `eds.Spec.Scaling.MinIndexReplicas = 0` to this test instead of changing the index replicas to not alter the actual test spec.

File: operator/autoscaler_test.go [hunk 2]
```diff
@@ -341,15 +341,36 @@ func TestScaleUpCausedByShardToNodeRatioExceeded(t *testing.T) {
...
 	eds := edsTestFixture(4)
 	esNodes := make([]ESNode, 0)
 
 	// don't scale down if that would violate the maxShardsPerNode
-	eds.Spec.Scaling.MaxShardsPerNode = 6
+	eds.Spec.Scaling.MaxShardsPerNode = 12
 	esIndices := map[string]ESIndex{
-		"ad1": {Replicas: 1, Primaries: 9, Index: "ad1"},
-		"ad2": {Replicas: 0, Primaries: 5, Index: "ad2"},
+		"ad1": {Replicas: 1, Primaries: 18, Index: "ad1"},
```
Comment (otrosien) [inline operator/autoscaler_test.go:372]: this test unveils an interesting difference in behaviour that we should capture as test. Previously it seems that scaling down didn't align the index replicas like it does now.

File: operator/autoscaler.go [hunk 3]
```diff
@@ -283,12 +300,22 @@ func (as *AutoScaler) scaleUpOrDown(esIndices map[string]ESIndex, scalingHint Sc
 				})
 			}
 			if newTotalShards != currentTotalShards {
-				newDesiredNodeReplicas := as.ensureUpperBoundNodeReplicas(scalingSpec, calculateNodesWithSameShardToNodeRatio(currentDesiredNodeReplicas, currentTotalShards, newTotalShards))
+				newDesiredNodeReplicas := currentDesiredNodeReplicas
+
+				currentShardToNodeRatio = float64(newTotalShards) / float64(currentDesiredNodeReplicas)
```
Comment (otrosien) [inline operator/autoscaler.go:305]: what you're calculating here is not the current shard-to-node ratio, it is actually the target shard-to-node ratio if we kept the same number of nodes. Therefore the log message below "Keeping shard-to-node ratio (%.2f), and increasing index replicas." would be wrong.

File: operator/autoscaler_test.go (hunk 2 above)
Comment (s-vkropotko) [reply operator/autoscaler_test.go:372]: I think I've already added test for this
https://github.com/zalando-incubator/es-operator/pull/157/files#diff-79598d30b7d766804171bf9a57fb7b2833b1f3e3ca7073e083d43fe8417477d4R385

ScalingHint=Down but we expect increasing number of replicas and an actual behavior is scaling UP

File: operator/autoscaler.go (hunk 3 above)
Comment (s-vkropotko) [reply operator/autoscaler.go:305]: yes, it's a good catch. I will fix it

File: operator/autoscaler_test.go [hunk 4]
```diff
@@ -341,15 +341,36 @@ func TestScaleUpCausedByShardToNodeRatioExceeded(t *testing.T) {
 	require.Equal(t, UP, actual.ScalingDirection, actual.Description)
 }
 
+func TestScaleDownCausedByShardToNodeRatioLessThanOne(t *testing.T) {
+	eds := edsTestFixture(12)
+	esNodes := make([]ESNode, 0)
+
+	// require to scale-up because we exceeded shard-to-node ratio limits.
```
Comment (otrosien) [inline operator/autoscaler_test.go:348]: change to 'require to scale-up index replicas because we are below one shard per node.'

File: operator/autoscaler_test.go (hunk 2 above)
Comment (otrosien) [reply operator/autoscaler_test.go:372]: Here we have an edge-case situation of index replicas being different within the EDS. I think this deserves its own test. The other test is fine. 

Can you also add a comment with the math why we'd be exceeding max-shards per node?

File: operator/autoscaler.go [hunk 5]
```diff
@@ -283,12 +300,22 @@ func (as *AutoScaler) scaleUpOrDown(esIndices map[string]ESIndex, scalingHint Sc
...
+				// Evaluate new number of nodes only if we above MinShardsPerNode parameter
+				if currentShardToNodeRatio >= float64(scalingSpec.MinShardsPerNode) {
```
[the remark this answers is not in the document]
Comment (s-vkropotko) [reply operator/autoscaler.go]: https://github.com/zalando-incubator/es-operator/pull/157/files#diff-79598d30b7d766804171bf9a57fb7b2833b1f3e3ca7073e083d43fe8417477d4R345

File: operator/autoscaler_test.go (hunk 2 above)
Comment (s-vkropotko) [reply operator/autoscaler_test.go:374]: > Here we have an edge-case situation of index replicas being different within the EDS. I think this deserves its own test. The other test is fine.

Sorry, I didn't get this. Could you please explain it or provide some example?

File: operator/autoscaler_test.go [hunk 6]
```diff
@@ -341,16 +342,46 @@ func TestScaleUpCausedByShardToNodeRatioExceeded(t *testing.T) {
 	require.Equal(t, UP, actual.ScalingDirection, actual.Description)
 }
 
+func TestScaleDownCausedByShardToNodeRatioLessThanOne(t *testing.T) {
```
Comment (otrosien) [inline operator/autoscaler_test.go:345]: nit: should be "TestScaleUp"

File: operator/autoscaler.go (hunk 3 above)
Comment (otrosien) [reply operator/autoscaler.go]: 👍

File: operator/autoscaler.go [hunk 7]
```diff
@@ -225,46 +225,59 @@ func (as *AutoScaler) getMaxDiskUsage(managedNodes []ESNode) float64 {
...
-
-func (as *AutoScaler) ensureUpperBoundNodeReplicas(scalingSpec *zv1.ElasticsearchDataSetScaling, newDesiredNodeReplicas int32) int32 {
+func (as *AutoScaler) ensureBoundsNodeReplicas(scalingSpec *zv1.ElasticsearchDataSetScaling, newDesiredNodeReplicas int32) int32 {
 	if scalingSpec.MaxReplicas > 0 {
-		upperBound := int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
-		if upperBound < newDesiredNodeReplicas {
+		newDesiredNodeReplicas = int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
+		if int32(scalingSpec.MaxReplicas) < newDesiredNodeReplicas {
 			as.logger.Warnf("Requested to scale up to %d, which is beyond the defined maxReplicas of %d.", newDesiredNodeReplicas, scalingSpec.MaxReplicas)
+			return newDesiredNodeReplicas
```
Comment (otrosien) [inline operator/autoscaler.go:233]: This looks wrong to me. Are we going to allow scaling beyond scalingSpec.MaxReplicas?

Comment (s-vkropotko) [reply operator/autoscaler.go:233]: I think it's just a mistake and the total logic in this function is not correct
```
		newDesiredNodeReplicas = int32(math.Min(float64(newDesiredNodeReplicas), float64(scalingSpec.MaxReplicas)))
		if int32(scalingSpec.MaxReplicas) < newDesiredNodeReplicas {
			as.logger.Warnf("Requested to scale up to %d, which is beyond the defined maxReplicas of %d.", newDesiredNodeReplicas, scalingSpec.MaxReplicas)
			return newDesiredNodeReplicas
		}
```
I will update it

Comment (s-vkropotko) [reply operator/autoscaler.go]: I updated this function and add a test for it
https://github.com/zalando-incubator/es-operator/pull/157/files#diff-add97f45040e000fdf77c6cab0697b142d788e56b1379176785662a659a5186bR228

File: operator/autoscaler_test.go (hunk 2 above)
Comment (otrosien) [reply operator/autoscaler_test.go:374]: all good, I think we got all covered now.

**Comment by otrosien** [comment]
👍

**Comment by mikkeloscar** [comment]
:+1:
